### ISSUE TITLE
no such file or directory: ~/.zshrc

### ISSUE THREAD

**Issue by simonszu** [jandamm/zgenom#92]
Hi, 
i have my zgenom data in `~/.zsh/zgenom`. This file is sourced in the .zshrc like this: 

```
# Load zgenom stuff. Needs to stay at top so it can output stuff before the
# Powerlevel10k instant prompt can come into place
source $HOME/.zsh/zgenom
```

I get an error each time i do a `zgenom reset` and reinitializing zgenom: 

```
-- zgenom: Updating 'zsh-users/zsh-completions/___' ...
From https://github.com/zsh-users/zsh-completions
   bebaa61..a1a11ca  master     -> origin/master
Updating bebaa61..a1a11ca
Fast-forward
 src/_ccache    | 2 +-
 src/_drush     | 4 ++--
 src/_golang    | 2 +-
 src/_jonas     | 2 +-
 src/_openssl   | 4 ++--
 src/_perf      | 2 +-
 src/_pm2       | 4 ++--
 src/_scala     | 2 +-
 src/_tmuxp     | 2 +-
 src/_udisksctl | 2 +-
 src/_yarn      | 2 +-
 11 files changed, 14 insertions(+), 14 deletions(-)

-- zgenom: Deleting `/Users/simonszu/.zgenom/sources/init.zsh` ...

Recreating init.zsh

zsh:source:1: no such file or directory: ~/.zshrc
--------
```

Why that? The file is there and can immediately sourced manually after the init process finishes and i get my prompt back.

**Comment by jandamm** [comment]
When you're executing `zgenom reset` and restart zsh the plugins aren't actually updated - only new ones are installed.

The `Recreating init.zsh` hints that you're using `zgenom autoupdate`.
It might be an issue because the init.zsh doesn't exist while running autoupdate.

Does this happen every time you run `zgenom reset` and restart zsh?

Would you mind replacing line 83 of `functions/zgenom-autoupdate` with
```
&& printf 'Recreating init.zsh\n\n' && echo ~ \
```
and then check if the output matches what you get when running `echo ~` in the interactive shell?
To force the code to be executed you have to remove `$ZGEN_DIR/.zgenom-plugin-lastupdate`.

**Comment by simonszu** [comment]
This happens every time a `zgenom update` is called (after a restart of zsh), but not when `zgenom reset` is executed. 

I have modified the file, and the output is the same as `echo ~`.

**Comment by jandamm** [comment]
Hey @simonszu, thanks for your reply.

I only got more questions... 😅

Do you mean `autoupdate` instead of `update`?
Because `zgenom update` shouldn't source your zshrc.

Does `cat $ZGEN_DIR/.zgenom-plugin-lastupdate` succeed and print an integer?

Does it work with `zgenom autoupdate --no-background`?

Does that happen in every new shell or only after some days?

Do you have your dotfiles available online or could you share `$HOME/.zsh/zgenom` (you can redact the plugins if this is a concern of yours).

**Comment by jandamm** [comment]
Could you recreate your alias using absolute paths?
Someone at StackOverflow mentioned this message would appear sometimes when relative paths are used for the symlinks.
https://stackoverflow.com/a/66049244

`ln -s $HOME/path/to/zshrc $HOME/.zshrc`

**Comment by simonszu** [comment]
Hi @jandamm sorry for the late reply. Here are my answers: 

> Do you mean autoupdate instead of update?
> Because zgenom update shouldn't source your zshrc.

I did a `zgenom update` which deleted my `init.zsh`. After that i restarted the shell, and the error occured. However, interestingly, i tried it once again today, and the issue did not happen. I am currently unsure what could have changed. 
But i have noticed that the issue occurs (if it occurs) directly after i start a new terminal session (which will source zgenom) and i suppose that's the autoupdate, since it checked plugins for new releases. 

> Does cat $ZGEN_DIR/.zgenom-plugin-lastupdate succeed and print an integer?

Yes

> Does it work with zgenom autoupdate --no-background?

Yes, currently it does nothing, i suppose because my plugins are at the newest state. 

> Does zsh -c source ~/.zshrc work when executed in the terminal?

No. `zsh:source:1: not enough arguments`. But calling the `source` directly in the current shell works. 

> Does it always show the exact same message including the updates or do the updates differ?

I have to copy-paste the exact output at least two times and comparing it. Give me some time to do that, since: 

> Does that happen in every new shell or only after some days?

Only after some days, after it has downloaded a new version of a plugin. 

> Do you have your dotfiles available online or could you share $HOME/.zsh/zgenom (you can redact the plugins if this is a concern of yours).

I don't have them publicly available, since there is also my `~/.ssh` folder in them, but here you are: 

https://pastebin.com/W4MmHNL6

It contains some logic to automatically install zgenom on a new machine where i have cloned my dotfiles onto, but that should not interfere with zgenom itself. 

> Could you recreate your symlink using absolute paths?

My `.zshrc` is not a symlink. I manage my dotfiles with yadm, and it basically just wraps a git around my $HOME with an implicit gitignore-exclude-all, so my zshrc is a real file: 

```
-rw-r--r--  1 simonszu simonszu  890 Nov 24 13:25 .zshrc
```

**Comment by jandamm** [comment]
Thanks @simonszu.
So the issue is in `zgenom autoupdate` but I'm not yet sure why it doesn't work.

Could you please check `zsh -c "source ~/.zshrc"` (I forgot the quotes last time...)

What happens when you
```
rm $ZGEN_DIR/.zgenom-system-update && zsh -c exit
# wait some time
zsh
```
This should give you a log like the one you posted.
- Does the error still exist?
- What happens when you add `--no-background` to autoupdate in ~/.zsh/zgenom and retry?
- Could you replace the following line with `local cmd="source /Users/simonszu/.zshrc" and retry?
https://github.com/jandamm/zgenom/blob/fc5f715123b779ad63f6e56cf01c59ddba4d9692/functions/zgenom-autoupdate#L79

**Comment by simonszu** [comment]
Hi, 
here are my results: 

> Could you please check zsh -c "source ~/.zshrc" (I forgot the quotes last time...)

From a visual perspective it just seems that this prompt does exactly nothing. So it seems to source the .zshrc quite well. 



> rm $ZGEN_DIR/.zgenom-system-update && zsh -c exit

This yields an interesting result: 
`rm: das Entfernen von '/home/simonszu/.zgenom/sources/.zgenom-system-update' ist nicht möglich: Datei oder Verzeichnis nicht gefunden` (File or directory not found) - but i have a `.zgenom-system-lastupdate`. What about that? 

Also i have noticed that although `zgenom autoupdate` did output nothing, an immediately executed `zgenom update` afterwards updated some plugins and finished with `-- zgenom: Deleting `/home/simonszu/.zgenom/sources/init.zsh` ...
` Don't know if it is relevant. 

> My assumption is that somehow ~ isn't expanded to /Users/simonszu as it should be.

Just a side information: I have experienced the issue at first in zsh in WSL, on a debian base with zsh 5.8, and also on a native Linux box with Debian, also zsh 5.8. You are apparently using a mac (which i happen to own as well), but it is zsh 5.8 as well, installed via homebrew. But interestingly i do not remember having experienced the issue there.

**Comment by jandamm** [comment]
> rm: das Entfernen von '/home/simonszu/.zgenom/sources/.zgenom-system-update' ist nicht möglich: Datei oder Verzeichnis nicht gefunden (File or directory not found) - but i have a .zgenom-system-lastupdate. What about that?

Right, could you retry with `.zgenom-system-lastupdate` instead?

> Also i have noticed that although zgenom autoupdate did output nothing
This is correct. Autoupdate will dispatch a task in the background which writes to a file and the file is then read when the next shell is started. That behavior should be changed by adding `--no-background`.

Could you please retest with this script
```
rm $ZGEN_DIR/.zgenom-system-lastupdate && zsh -c exit
# wait some time
zsh
```

1. Without changes
2. What happens when you add --no-background to autoupdate in ~/.zsh/zgenom
3. Could you replace the following line with local cmd="source /Users/simonszu/.zshrc"
https://github.com/jandamm/zgenom/blob/fc5f715123b779ad63f6e56cf01c59ddba4d9692/functions/zgenom-autoupdate#L79

I'd hope that at least 3. fixes the issue.

I've got the detection of the `.zshrc` path from https://github.com/romkatv/powerlevel10k/blob/f217e4a39a284f6db7be7a4cfde8647085f97865/gitstatus/gitstatus.plugin.zsh#L794.
From my understand the `~` shouldn't be necessary and a absolute path could be used..

**Comment by simonszu** [comment]
OK, so: 

> Without changes

Error occurs. 

> What happens when you add --no-background to autoupdate in ~/.zsh/zgenom

Error occurs.

> Could you replace the following line with local cmd="source /Users/simonszu/.zshrc"

Error does **not** occur. 

So the root cause is found. I am not so fluent in shell scripting to understand the deeper details of all the magic stuff in this command in line 79, but i will be happy to help you with fixation attempts to make it more universal.

**Comment by jandamm** [comment]
I probably will just change the command from `source ~/.zshrc` to `source /Users/simonszu/.zshrc` since this seems to work without issues.

Would be nice if you could change line 79 to `local cmd="echo ${${${(q)${ZDOTDIR:-~}}/#${(q)home}/'~'}//\%/%%}/.zshrc"` and retry.

**Comment by simonszu** [comment]
Sure. 

```
Already up to date.

-- zgenom: Deleting `/home/simonszu/.zgenom/sources/init.zsh` ...

Creating a zgenom save
-- zgenom: Creating `/home/simonszu/.zgenom/sources/init.zsh` ...
-- zgenom: Initializing completions ...
-- zgenom: Compiling files ...
╭─ ~ ································································································································································· ✔  15:18:01
╰─
```
Out of curiosity i executed this echo command in an interactive session as well. There it works: 
```
❯ echo ${${${(q)${ZDOTDIR:-~}}/#${(q)home}/'~'}//\%/%%}/.zshrc
/home/simonszu/.zshrc
```

**Comment by jandamm** [comment]
@simonszu I hope it's fixed in #95.
Could you check the branch out at retry these steps:

```
rm $ZGEN_DIR/.zgenom-system-lastupdate && zsh -c exit
# wait some time
zsh
```

If everything is alright I'll merge it 👍

**Comment by simonszu** [comment]
@jandamm looks good: 
```
-- zgenom: Deleting `/home/simonszu/.zgenom/sources/init.zsh` ...

Creating a zgenom save
-- zgenom: Creating `/home/simonszu/.zgenom/sources/init.zsh` ...
-- zgenom: Initializing completions ...
-- zgenom: Compiling files ...

╭─ ~ ································································································································································· ✔  09:21:46
╰─
```
